hasktorch/src/Torch/Tensor.hs

```diff
@@ -218,20 +238,112 @@ toCUDA t = unsafePerformIO $ (cast1 ATen.tensor_cuda) t
 -- Indexing support
 --------------------------------------------------------------------------------
 
+-- TensorIndex is the same as slice of pytorch.
+--
+-- There is one-to-one correspondence between Pytorch and Hasktorch tensor index types:
+-- Pytorch                 | Hasktorch
+-- -----------------------------------------------------
+-- `None`                  | `None`
+-- `Ellipsis`              | `Ellipsis`
+-- `...`                   | `Ellipsis`
+-- `123`                   | `123`
+-- `True` / `False`        | `True` / `False`
+-- `:`                     | `Slice ()`
+-- `::`                    | `Slice ()`
+-- `1:`                    | `Slice (1, None)`
+-- `1::`                   | `Slice (1, None)`
+-- `:3`                    | `Slice (None, 3)`
+-- `:3:`                   | `Slice (None, 3)`
+-- `::2`                   | `Slice (None, None, 2)`
+-- `1:3`                   | `Slice (1, 3)`
+-- `1::2`                  | `Slice (1, None, 2)`
+-- `:3:2`                  | `Slice (None, 3, 2)`
+-- `1:3:2`                 | `Slice (1, 3, 2)`
+-- `torch.tensor([1, 2])`) | `asTensor([1, 2 ::Int])`
+
+
 (@@) :: TensorIndex a => Tensor -> a -> Tensor
 t @@ idx = fst $ indexWith (t, 0) idx
 
+data None = None
+  deriving (Show, Eq)
+
+data Ellipsis = Ellipsis
+  deriving (Show, Eq)
+
+data Slice a = Slice a
+  deriving (Show, Eq)
+
 class TensorIndex a where
   indexWith :: (Tensor, Int) -> a -> (Tensor, Int)
 
+
+-- ToDo:
+--   1, offset may have a bug.
+--   2, Implement Ellipsis and Boolean
+
+instance TensorIndex (Slice (Integer,Integer)) where
+  indexWith (t, offset) (Slice (start,end)) = (slice t offset (fromIntegral start) (fromIntegral end) 1, offset+1)
+
+instance TensorIndex (Slice (Int,Int)) where
```

Comment:
I think you can use the appropriate integral type class to save on repetition

Reply:
I've tried integral type class. 
```
instance {-# OVERLAPS #-}(Integral a) => TensorIndex (Slice (a,a)) where
  indexWith (t, offset) (Slice (start,end)) = (slice t offset (fromIntegral start) (fromIntegral end) 1, offset+1)
```
But I've met following errors.

![image](https://user-images.githubusercontent.com/2469618/86077238-94be1e80-bac6-11ea-869d-b27154f3abd2.png)